Add IBMCA Provider for OpenSSL 3.0

ifranzki opened the pull request:
Things to discuss:

- Should the provider have the same version number than the Engine? Currently it has 1.1.0 since it is the first version of the provider, but the engine has 2.2.2....
- Should we maybe create a new release (i.e. 2.3.0) for this (or even a version - 3.0.0.)?

ifranzki commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -30,8 +31,97 @@ specify:
+.B $IBMCA_DEBUG
+is set to \fBstderr\fP debug output to stderr for the IBMCA provider is enabled.
+If it is set to \fBon\fP the debug output is written into a trace file in
+\fB[/usr/local]/var/log/ibmca/trace.<pid>\fP, where <pid> is the process ID of
```
This should be [`/usr/local]/var/log/ibmca/trace-<provider-name>.<pid>` like above.
Will fix this but wait with force-push for more updates.

ifranzki commented on README.md:
```diff
@@ -1,14 +1,16 @@
  * openssl-devel >= 0.9.8
+ * openssl-devel >= 3.0.0 for building the IBMCA provider
  * libica-devel >= 3.3.0
+ * libica-devel >= 3.6.0 of >= 4.0.0 for building the IBMCA provider
```
I guess this should be 4.0.1 if Joerg creates a bugfix release for libica

ifranzki replied:
and 'or' instead of 'of'

ifranzki commented on README.md:
```diff
@@ -17,42 +19,77 @@ The build requirements are:
  * openssl >= 0.9.8
+ * openssl-devel >= 3.0.0 for using the IBMCA provider
  * libica >= 3.3.0
+ * libica >= 3.6.0 of >= 4.0.0 for using the IBMCA provider
```
same here

jschmidb commented on configure.ac:
```diff
@@ -90,6 +90,12 @@ fi
 
+AC_ARG_WITH([provider-libica-full],
+	[AS_HELP_STRING([--with-provider-libica-full],[Use the full libica as library for the IBMCA provider])],
+	[useproviverfulllibica=${withval}],
```
useproviverfulllibica -> useproviderfulllibica

ifranzki replied:
fixed

jschmidb commented on configure.ac:
```diff
@@ -90,6 +90,12 @@ fi
+	[AS_HELP_STRING([--with-provider-libica-full],[Use the full libica as library for the IBMCA provider])],
+	[useproviverfulllibica=${withval}],
+	[])
+AM_CONDITIONAL([PROVIDER_FULL_LIBICA], [test "x$useproviverfulllibica" = xyes])
```
useproviverfulllibica -> useproviderfulllibica

jschmidb commented:
commit title: Prepare provider for implementing cryto operations  -> crypto

jschmidb commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -20,7 +20,8 @@ There is no need to use another provider for accelerating those.
 This section includes the identity (i.e. name) of the provider (i.e. ibmca),
 the module implementing the provider (ibmca-provider.so), provider parameters,
-and if it is to be actiavated automatically.
+and if it is to be actiavated automatically. For a description of the
```
actiavated -> activated

ifranzki replied:
fixed

jschmidb commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -30,8 +31,97 @@ specify:
+.RS
+Enables FIPS mode for the IBMCA provider. If FIPS is enabled, it is ensured that
+the \fBlibica-cex\fP library is built with FIPS enabled, and the system is
+running in FIPS mode. If thats not the case, the IBMCA provider initialization
```
that's

ifranzki replied:
fixed

jschmidb commented on test/provider/threadtest.c:
```diff
@@ -0,0 +1,217 @@
+    pthread_t me;
+    EVP_PKEY *eckey = NULL;
+    EVP_PKEY_CTX *pctx = NULL;
+    
```
line 116: white spaces

ifranzki replied:
fixed

jschmidb commented on test/provider/threadtest.c:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    setup();
+    
```
line 156: white spaces

ifranzki commented on configure.ac:
```diff
@@ -138,3 +144,8 @@ echo "CFLAGS=$CFLAGS"
 echo "IBMCA engine:      $enable_engine"
 echo "  default library: $defaultlib"
 echo "IBMCA provider:    $enable_provider"
+if test "x$useproviverfulllibica" = xyes; then
```
fixed

ifranzki commented on test/provider/threadtest.c:
```diff
@@ -0,0 +1,217 @@
+    while (1) {
+        int option_index;
+        static struct option long_options[] =
+            {
+             {"threads", required_argument, 0, 't'},
+             {0,         0,                 0, 0  }
+            };
```
fixed

ifranzki commented:
> commit title: Prepare provider for implementing cryto operations -> crypto

fixed

juergenchrist reviewed (changes requested):
In general, the Copyright year is everywhere 2021.  Should include 2022.

ifranzki commented:
>In general, the Copyright year is everywhere 2021. Should include 2022

Fixed, did not think about updating the Copyright year when the new year stated...

ifranzki commented on openssl-ibmca-provider.spec:
```diff
@@ -0,0 +1,112 @@
+
+%description
+This package contains a shared object OpenSSL dynamic provider which interfaces 
+to libica, a library enabling the IBM s390/x CPACF crypto instructions.
```
fixed

ifranzki commented on src/provider/Makefile.am:
```diff
@@ -0,0 +1,32 @@
+noinst_SCRIPTS = ibmca-provider-opensslconfig
+
+install-data-hook:
+	$(MKDIR_P) $(logdir)
```
Are you sure? in configure.ac:
`logdir=$localstatedir/log/ibmca`
and $localstatedir already contains the DESTDIR, no?

juergenchrist replied:
No, nothing contains DESTDIR.  This is purely used in staged installations, so not contained in any other variable, but to be used during every target run during installation.  See https://www.gnu.org/prep/standards/standards.html#DESTDIR.

ifranzki replied:
ok, added.

ifranzki commented on openssl-ibmca-provider.spec:
```diff
@@ -0,0 +1,112 @@
+rm -f $RPM_BUILD_ROOT%{modulesdir}/ibmca-provider.la
+mv -f src/provider/openssl.cnf.sample src/provider/openssl.cnf.sample.%{_arch}
+
+%files
```
fixed

ifranzki commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -0,0 +1,127 @@
+.\" Process this file with
```
Certainly right, but we don't have that in the engine man page either. I leave it to you to update the engines man page later on.

ifranzki commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -0,0 +1,127 @@
+the process ID of the current process.
+.PP
+.SH SEE ALSO
+.B provider(1)
```
added

ifranzki commented on src/provider/ibmca-provider-opensslconfig:
```diff
@@ -0,0 +1,137 @@
+# USE WITH CARE: No automation can replace human knowledge and understanding
+#                of the desired configuration!
+#
+
```
added

ifranzki commented on test/provider/threadtest.c:
```diff
@@ -0,0 +1,216 @@
+#include <openssl/provider.h>
+#include <openssl/err.h>
+
+/* This is just a random number of threads to stimulate engine configuration. */
```
changed

ifranzki commented on test/provider/threadtest.c:
```diff
@@ -0,0 +1,216 @@
+                           CONF_MFLAGS_IGNORE_MISSING_FILE);
+}
+
+static int check_globals(EVP_PKEY *eckey)
```
The test creates one key outside of the treads and uses it in all of the threads. This causes the fallback-PKEY to be created inside the threads concurrently, and thus test locking of the fallback-PKEY-cache. So it tests thread safety of per key 'globals'.

OK, we could name the function something else than check_globals() of course....

juergenchrist replied:
Okay, I did not see this.  No need to rename this function then.

ifranzki commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -0,0 +1,127 @@
+module =
+.I ibmca-provider.so
+.RS
+Set the name and optionally the path to the IBMCA provider shared object file
```
added some text about MODILESDIR and OPENSSL_MODULES

ifranzki commented on src/provider/p_ibmca.c:
```diff
@@ -0,0 +1,1392 @@
+        return;
+
+    pthread_mutex_lock(&provctx->debug_mutex);
+    /* no error checking here: if lock fails print trace msg anyway */
```
done

ifranzki commented on src/provider/p_ibmca.c:
```diff
@@ -0,0 +1,1392 @@
+    ibmca_debug_ctx(provctx, "provctx: %p", provctx);
+
+    for (num = 0; ica_mech_infos[num].algo != NULL; num++)
+        ;
```
Yes, but I would like to keep this a NULL-terminated list anyway. It might happen in the future that ica_mech_infos becomes an external variable and then the size would not be to determine that easy outside of this module. For consistency, I would like to keep the size determination as is also here.

ifranzki commented on src/provider/p_context.c:
```diff
@@ -0,0 +1,425 @@
+        ibmca_keymgmt_free(ctx->key);
+
+    if (ctx->propq != NULL)
+        P_FREE(ctx->provctx, (void *)ctx->propq);
```
changed

ifranzki commented on src/provider/rsa_keymgmt.c:
```diff
@@ -0,0 +1,1868 @@
+    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
+        ok = ok && (key1->rsa.private.key_length ==
+                           key2->rsa.private.key_length &&
+                    memcmp(key1->rsa.private.p,
```
fixed

ifranzki commented on src/provider/rsa_padding.c:
```diff
@@ -0,0 +1,1173 @@
+        error++;
+
+    ps_len = oaep_md_len;
+    while ((dbmask[ps_len] == 0x00) && (ps_len < dbmask_len))
```
good catch. swapped.

ifranzki commented on src/provider/ec_keymgmt.c:
```diff
@@ -0,0 +1,2424 @@
+
+    for (i = 0, all_zero = true; rc != EINVAL && i < len && all_zero; i++) {
+        if (q[i] != 0)
+            all_zero = false;
```
right, added break

ifranzki commented on src/provider/ec_keymgmt.c:
```diff
@@ -0,0 +1,2424 @@
+        goto out;
+    }
+
+    for (i = 0, all_zero = true; rc != EINVAL && i < len && all_zero; i++) {
```
ok, checked for rc == EINVAL before loop

ifranzki commented on src/provider/ec_keymgmt.c:
```diff
@@ -0,0 +1,2424 @@
+    }
+
+    for (i = 0, all_zero = true; rc != EINVAL && i < len && all_zero; i++) {
+        if ((*d)[i] != 0)
```
changed

ifranzki commented on src/provider/ec_keymgmt.c:
```diff
@@ -0,0 +1,2424 @@
+        put_error_key(key, IBMCA_ERR_INTERNAL_ERROR,
+                      "EC_POINT_set_affine_coordinates failed");
+        goto out;
+    }
```
We are just returning the EC key components of the key here. We can't influence at this point where the EC key components came from, so a check here is useless. 

Also, there is a provider function to perform a key check  (ibmca_keymgmt_ec_validate). There is no need to check the key components at other places.

ifranzki commented on src/provider/ec_keyexch.c:
```diff
@@ -0,0 +1,750 @@
+static int ibmca_keyexch_ec_set_ctx_params(void *vctx,
+                                           const OSSL_PARAM params[]);
+
+static void*ibmca_keyexch_ec_newctx(void *vprovctx)
```
fixed

ifranzki commented on src/provider/ec_keyexch.c:
```diff
@@ -0,0 +1,750 @@
+    *secretlen = outlen < ctx->key->ec.prime_size ?
+                                            outlen : ctx->key->ec.prime_size;
+
+    if (outlen < *secretlen) {
```
removed the if

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    ibmca_debug_key(key, "key: %p", key);
+
+    if (key->dh.priv != NULL)
+        BN_free(key->dh.priv);
```
changed

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    if (n == 0)
+        n = -1;
+
+    if (key->dh.ffc_params.p != NULL)
```
Well, this is what OpenSSL does:

```
int DH_security_bits(const DH *dh)
{
    int N;

    if (dh->params.q != NULL)
        N = BN_num_bits(dh->params.q);
    else if (dh->length)
        N = dh->length;
    else
        N = -1;
    if (dh->params.p != NULL)
        return BN_security_bits(BN_num_bits(dh->params.p), N);
    return -1;
}
```

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+        if (BN_is_word(key->dh.ffc_params.g, DH_GENERATOR_2) &&
+            !BN_is_bit_set(key->dh.ffc_params.p, 2)) {
+             /* clear bit 0, since it won't be a secret anyway */
+             if (BN_clear_bit(key->dh.priv, 0) != 1) {
```
Hm, well, the BN_clear_bit implementation does have some error paths.  I would not want to optimize this code to the current implementation of BN_clear_bit....
Does it really hurt to have this error code path, even if it is never hit ?

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    } else {
+        if (ibmca_keymgmt_dh_gen_priv_key(key, bn_ctx,
+                                          BN_num_bits(key->dh.ffc_params.q),
+                                          key->provctx->fips ? 112 : 80) != 1) {
```
From OpenSSL's DH keygen implementation in crypto/dh/dh_key.c

```
#ifdef FIPS_MODULE
# define MIN_STRENGTH 112
#else
# define MIN_STRENGTH 80
#endif

                if (!ossl_ffc_generate_private_key(ctx, &dh->params,
                                                   BN_num_bits(dh->params.q),
                                                   MIN_STRENGTH,
                                                   priv_key))
```

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+        }
+    } else {
+        put_error_key(key, IBMCA_ERR_INTERNAL_ERROR,
+                      "Neither parmgen context nor template key available");
```
fixed

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    if (ibmca_keymgmt_match(key1, key2) == 0)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
```
I would agree to that, but OpenSSL just changed their logic in the default provider with the following commit:
https://github.com/openssl/openssl/commit/ee22a3741e3fc27c981e7f7e9bcb8d3342b0c65a

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```
no, OSSL_KEYMGMT_SELECT_KEYPAIR is a mask out of 2 bits:
```
# define OSSL_KEYMGMT_SELECT_KEYPAIR            \
    ( OSSL_KEYMGMT_SELECT_PRIVATE_KEY | OSSL_KEYMGMT_SELECT_PUBLIC_KEY )
```
So it can happen that only one of the bit is on, but the other is off.

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+            }
+        }
+
+        if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
```
see above

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    if (rc == 0)
+        return 0;
+    if (rc > 0) {
+        key->dh.pub = BN_bin2bn(enc, size, key->dh.pub);
```
good catch. added a BN_free in case key->dh.pub is not NULL.

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+        /* public key must be same size as p */
+        if (key->dh.pub == NULL ||
+            BN_is_zero(key->dh.pub) ||
+            size != ibmca_keymgmt_dh_get_prime_size(key)) {
```
good catch. add a BN_free and key->dh.pub = NULL in the error case.

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+#define IBMCA_TLS_GROUP_ID_ffdhe6144        259
+#define IBMCA_TLS_GROUP_ID_ffdhe8192        260
+
+static const struct ibmca_tls_group_constants ibmca_tls_group_consts[8] = {
```
copy/paste error. fixed.

ifranzki commented on src/provider/dh_keyexch.c:
```diff
@@ -0,0 +1,813 @@
+    ctx->dh.derive.kdf_ukm = NULL;
+    ctx->dh.derive.kdf_ukmlen = 0;
+
+    if (ctx->dh.derive.kdf_ukm != NULL)
```
yea, left over copy.... removed

ifranzki commented on src/provider/dh_keyexch.c:
```diff
@@ -0,0 +1,813 @@
+        return 0;
+    }
+
+    /* Setup defaults for this context */
```
changed

ifranzki commented on src/provider/dh_keyexch.c:
```diff
@@ -0,0 +1,813 @@
+
+    *secretlen = outlen < prime_size ? outlen : prime_size;
+
+    if (outlen < *secretlen) {
```
removed if

ifranzki commented on src/provider/dh_keyexch.c:
```diff
@@ -0,0 +1,813 @@
+    if (use_tbuf)
+        memcpy(secret, ctx->tbuf, *secretlen);
+
+    if (ctx->dh.derive.pad == false) {
```
At this point secret is always filled with a padded secret. If no padding is wanted, we remove the leading zero bytes and thus un-pad it. So no padding case needed.

ifranzki commented on src/provider/dh_keyexch.c:
```diff
@@ -0,0 +1,813 @@
+
+    rc = ica_rsa_mod_expo(ctx->provctx->ica_adapter, pub, &mod_exp, buf);
+    if (rc == 0) {
+        z = BN_bin2bn(buf, prime_size, BN_secure_new());
```
fixed

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+    OSSL_PARAM_free(params);
+
+error:
+    OSSL_PARAM_BLD_free(bld);
```
Good catch. Fixed. Also for EC keymgmt.

ifranzki commented on src/provider/Makefile.am:
```diff
@@ -0,0 +1,32 @@
+
+install-data-hook:
+	$(MKDIR_P) $(logdir)
+	$(CHMOD) 0770 $(logdir)
```
added

juergenchrist commented on src/provider/p_key.c:
```diff
@@ -0,0 +1,493 @@
+{
+    ibmca_debug_key(key, "key: %p ", key);
+
+     if (key->ref_count > 0)
```
To be more specific:  The problem is that the value of the ref_count might be changed by another thread before we return from this function and decide based on the return value what to do.  The whole path from ref count decrement to the actual decision to delete the object has to be atomic to remove above race condition.

ifranzki replied:
You are right, the return of the new value is the real problem.

How about this:
```
unsigned int ibmca_keymgmt_downref(struct ibmca_key *key)
{
    const struct ibmca_prov_ctx *provctx = key->provctx;
    unsigned int cnt;

    ibmca_debug_key(key, "key: %p ", key);

    if (key->ref_count == 0)
        return 0;

    cnt = __sync_sub_and_fetch(&key->ref_count, 1);
    ibmca_debug_ctx(provctx, "ref_count: %u", cnt);

    return cnt;
}
```

With this, if 2 threads have a reference to the key (i.e. ref_count = 2) and then concurrently free the key, both find out that ref_counf is > 0 (no matter when they check), and then decrement the ref_count atomic. Each one returns its new value: One of it returns 1 and the other returns 0. Those that returned 0 will actually free the object. The other will not touch the key object anymore (not even for tracing).

What is still not thread save is the check for ref_count already being zero, but this really is a case where an error has happend before already. One must not have a reference to a key whos ref_count is zero. If you don't have a reference, you won't call downref(). Any change of the ref_count (which was zero) by another thread (which also would not be allowed to have a reference to the key with a zero ref_count) would be incorrect in itself.

juergenchrist replied:
Well, the `if (key->ref_count == 0)` case is anyway the case where somebody already violated our calling convention.  But the `return 0` would lead to a free on a key object that has already been freed and we do not get anything.  At least some debug message (maybe also an assertion?) and a non-0 return value such that we do not make things worse?  I simply do not like code paths that lead to double frees (even if someone violates our calling convention).

ifranzki replied:
Good point about the return 0..... How about this:
```
unsigned int ibmca_keymgmt_downref(struct ibmca_key *key)
{
    const struct ibmca_prov_ctx *provctx;
    unsigned int cnt;

    assert(key->ref_count > 0);

    provctx = key->provctx;
    ibmca_debug_ctx(provctx, "key: %p ", key);

    cnt = __sync_sub_and_fetch(&key->ref_count, 1);
    ibmca_debug_ctx(provctx, "ref_count: %u", cnt);

    return cnt;
}
```
It will abort with an assert if the ref_count is zero, and does not touch the key before the assertion. It still does not check for key being non-NULL, but that would not help much I guess anyway. If the key has already been freed, then the dereferencing of key to check the ref_count of course can already cause a segfault or similar....

juergenchrist replied:
The assert is only available if `NDEBUG` is not defined.  But that is fine for me.  It is anyway just a goody to try to figure out that we are already screwed...

ifranzki replied:
right. It could also be questioned if a library is supposed to do an abort (as part of the assert), but as you said, its just a try to make things not worse that they already are when the assert is hit.

ifranzki commented on src/provider/dh_keymgmt.c:
```diff
@@ -0,0 +1,2065 @@
+
+    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
+         /* OSSL_PKEY_PARAM_PRIV_KEY */
+        key->dh.priv = BN_secure_new();
```
added a NULL check to avoid non-secure BNs to be used in case of errors.

ifranzki commented on src/provider/p_ibmca.c:
```diff
@@ -0,0 +1,1392 @@
+     * libica performs FIPS self-tests if running in FIPS mode. If any of
+     * the fips tests failed, additional flags are on.
+     */
+    return (provctx->ica_fips_status == ICA_FIPS_MODE) ? 1 : 0;
```
Right, this is certainly not enough to claim FIPS compliance. May thinking was that we start with this minimal FIPS support, any add the full FIPS support including own selftests and HMAC checks in a later stage. Before we spend the effort for this I would like to know what the distros think about full FIPS support of the provider.

However, it is of course questionable if setting `fips=yes` for the property definitions of the IBMCA provider is correct with this minimal FIPS support.....

How about declaring this minimum FIPS support as an experimental 'feature' and making it compile-time dependent on `#ifdef IBMCA_FIPS` which can be defined at configure time ? That way we keep the code, but it won't be active by default. Once we add full FIPS support at a later time, we can easily enable it.

This would of course also affect other places, like the fips config option, man pages, documentation, etc. where the FIPS mode must be removed.

juergenchrist replied:
Personally I do not like build options.  Lets go the even simpler route and keep the code, but document in the man page that it is experimental and currently not FIPS certified.  I guess this will be enough and not introduce ugliness in the form of `#ifdef`s in the code.

ifranzki replied:
ok fine with me.

How about the following note in the man page at the end of the `fips = yes | no` configuration setting description:

_Note: The IBMCA provider is currently not FIPS certified. It does not itself perform any FIPS self tests nor  an  integrity check which would be required to be FIPS certifiable. It is only ensured that libica-cex library has successfully performed its self tests and integrity checks when FIPS mode is enabled.  If you do not  want  that  the  IBMCA provider  is used for property queries that include fips=yes, then disable FIPS mode of the IBMCA provider by setting fips = no in the provider configuration._

juergenchrist commented on src/provider/rsa_padding.c:
```diff
@@ -0,0 +1,1173 @@
+     */
+    if (ibmca_rsa_pkcs1_mgf1(provctx, seed_mask, oaep_md_len, masked_db,
+                             dbmask_len, mgf1_md) == 0)
+        error++;
```
I just thought since you already went part of the way...

ifranzki commented on src/provider/doc/ibmca-provider.man:
```diff
@@ -0,0 +1,127 @@
+and the system is running in FIPS mode. If both is true, then the FIPS mode is
+enabled for the IBMCA provider. If FIPS is enabled (either explicitly
+configured or due to auto-detection), the IBMCA provider registers its
+algorithm with \fBfips=yes\fP, thus the IBMCA algorithms may be fetched using
```
see comment on  your other comment regarding FIPS.

ifranzki commented on test/provider/test.pm:
```diff
@@ -0,0 +1,344 @@
+
+    my ($keylen, $tests, $max_file_size) = @_;
+
+    `$prov openssl list -providers | grep "name: ibmca"`;
```
changed this to use exit(99)

The engine tests use `exit(77) unless (`$eng openssl engine -c` =~ m/ibmca/);` to skip the tests if the engine is not loaded. Not sure if this is really a good idea ? Should the tests be skipped if the engine/provider is not loaded, or is this rather an error sutuation?

juergenchrist replied:
I consider this an error since `$eng` should make sure the engine is loaded.  If that is not the case, the engine failed to load and we have some error (not necessarily in the crypto operation, but in the engine itself).  I guess I have to clean up the test cases for the engine at some point.

ifranzki commented:
Made /var/og/ibmca 777 to allow anyone to write trace files.

ifranzki commented:
@juergenchrist I am ready with the changes. Please take a look again and let me know if there is anything else you want me to change.

juergenchrist reviewed (changes requested):
Some files (mostly test files) still miss a copyright header.  Could you please add this.

ifranzki commented:
> Some files (mostly test files) still miss a copyright header. Could you please add this.

Will do. You should then also update the Engine test files and add a copyright header there as well. 

> Now I see a copyright header, but with year 2018?

LOL copy & paste ... Will fix.

> Couldn't we alternatively specify the full path to the provider library? 

Thats what I wanted to say with  the sentence "Set the name **and optionally the path** to the IBMCA provider shared object file". I read somewhere the on some platforms the path can be specified here, but not on all platforms. I guess on Linux you can. 

> Also, the environment variable is only usable in non-setuid-root binaries.

Right, added "(ignored in set-user-ID and set-group-ID programs)" as it is stated in the OpenSSL man page for that environment variable.

juergenchrist reviewed (approved):
to be merged later once internal CI is set up

ifranzki commented:
Changed test.pm to not use fork() for the tls.pl testcase. Fork causes problems with closing file descriptors when running in the CI where stdin is a socket connection.

ifranzki commented:
rebased to current master

ifranzki commented:
Rebased again to current master.

Also fixed bug in DH keyexch related to a bug fixed in OpenSSL with PR https://github.com/openssl/openssl/pull/17859. Note that the `dhkey` testcase will fail until that OpenSSL PR is merged.
